oap-server/server-core/src/main/java/org/apache/skywalking/oap/server/core/profile/analyze/ProfileStackNode.java

```diff
@@ -0,0 +1,244 @@
+        if (codeSignature == null) {
+            codeSignature = stackList.get(0);
+        }
+        // add detected stack
+        this.detectedBy(stack);
+
+        // handle stack children
+        ProfileStackNode parent = this;
+        for (int depth = 1; depth < stackList.size(); depth++) {
+            String elementCodeSignature = stackList.get(depth);
+
+            // find same code signature children
+            ProfileStackNode childElement = null;
+            for (ProfileStackNode child : parent.children) {
+                if (Objects.equal(child.codeSignature, elementCodeSignature)) {
+                    childElement = child;
+                    break;
+                }
+            }
+
+            if (childElement != null) {
+                // add detected stack
+                childElement.detectedBy(stack);
+                parent = childElement;
+            } else {
+                // add children
+                ProfileStackNode childNode = newNode();
+                childNode.codeSignature = elementCodeSignature;
+                childNode.detectedBy(stack);
+
+                parent.children.add(childNode);
+                parent = childNode;
+            }
+        }
+    }
+
+    /**
+     * combine from other {@link ProfileStackNode}
+     * @param node
+     * @return
+     */
+    public ProfileStackNode combine(ProfileStackNode node) {
+        // combine this node
+        this.combineDetectedStacks(node);
+
+        // merge tree using LDR to traversal tree node
+        // using stack to avoid recursion
+        // merge key.children <- value.children
+        LinkedList<Pair<ProfileStackNode, ProfileStackNode>> stack = new LinkedList<>();
+        stack.add(new Pair<>(this, node));
+        while (!stack.isEmpty()) {
+            Pair<ProfileStackNode, ProfileStackNode> needCombineNode = stack.pop();
+
+            // merge value children to key
+            // add to stack if need to keep traversal
+            combineChildrenNodes(needCombineNode.key, needCombineNode.value, stack::add);
+        }
+
+        return this;
+    }
+
+    /**
+     * merge all children nodes to appoint node
+     * @param mergeTo
+     * @param combineNode
+     * @param needKeepEach if keep traversal, need report
+     */
+    private void combineChildrenNodes(ProfileStackNode mergeTo, ProfileStackNode combineNode, Consumer<Pair<ProfileStackNode, ProfileStackNode>> needKeepEach) {
+        for (ProfileStackNode childrenNode : combineNode.children) {
+            // find node from mergeTo children
+            ProfileStackNode combinedNode = mergeTo.children.stream().filter(n -> childrenNode.matches(n)).findFirst().orElse(null);
+
+            if (combinedNode == null) {
+                mergeTo.children.add(childrenNode);
+            } else {
+                combinedNode.combineDetectedStacks(childrenNode);
+                needKeepEach.accept(new Pair<>(combinedNode, childrenNode));
+            }
```

Comment:
You are iterating `combineNode` and process found or not, but
1. Once it is found, it has not been removed from `mergeTo`, then all elements all been process multiple times.
1. You seem missing node left in `mergeTo#children`, if some of them are not included in `combineNode`